Update scheduling docs per BZ

mburke5678 opened the pull request:
https://bugzilla.redhat.com/show_bug.cgi?id=1517586

ravisantoshgudimetla commented on admin_guide/scheduling/scheduler.adoc:
```diff
@@ -212,27 +213,40 @@ they have to be explicitly specified in the scheduler configuration file.
-. CheckNodeMemoryPressure
-. CheckNodeDiskPressure
+. `NoVolumeZoneConflict`. 
+. `MaxEBSVolumeCount`. Checks the maximum volume limit for a node.
```
Checks the maximum volume that can be attached to AWS instance.

ravisantoshgudimetla commented on admin_guide/scheduling/scheduler.adoc:
```diff
@@ -212,27 +213,40 @@ they have to be explicitly specified in the scheduler configuration file.
-. CheckNodeDiskPressure
+. `NoVolumeZoneConflict`. 
+. `MaxEBSVolumeCount`. Checks the maximum volume limit for a node.
+. `MaxGCEPDVolumeCount`. Checks the maximum PD volume limit for a node.
```
to a GCE instance

ravisantoshgudimetla commented on admin_guide/scheduling/scheduler.adoc:
```diff
@@ -212,27 +213,40 @@ they have to be explicitly specified in the scheduler configuration file.
+. `MaxGCEPDVolumeCount`. Checks the maximum PD volume limit for a node.
+. `MatchInterPodAffinity`. Checks if the pod affinity/antiaffinity rules permit the pod.
+. `NoDiskConflict`. Checks the volume requested by a pod is available.
+. `GeneralPredicates`
```
I think this could be further split into non-critical and essential. In fact, essential predicates are the ones through which resource requests go through to see if they fit.

mburke5678 replied:
@ravisantoshgudimetla Which ones are essential? Hard to tell from the code. 

This PR lists essential predicates as: PodFitsResources, PodFitsHostPorts, HostName, MatchNodeSelector, which we don't list here. 
https://github.com/kubernetes/kubernetes/pull/50990
Are we listing only non-critical here?

ravisantoshgudimetla commented on admin_guide/scheduling/scheduler.adoc:
```diff
@@ -212,27 +213,40 @@ they have to be explicitly specified in the scheduler configuration file.
+. `MatchInterPodAffinity`. Checks if the pod affinity/antiaffinity rules permit the pod.
+. `NoDiskConflict`. Checks the volume requested by a pod is available.
+. `GeneralPredicates`
+. `PodToleratesNodeTaints`. Checks if a pod tolertaions can tolerate the node taints.
```
pod can tolerate

ravisantoshgudimetla commented on admin_guide/scheduling/scheduler.adoc:
```diff
@@ -212,27 +213,40 @@ they have to be explicitly specified in the scheduler configuration file.
-. TaintTolerationPriority
+. `SelectorSpreadPriority`. The scheduler looks for services, RCs,RSs and StatefulSets that match the pod,
+then finds existing pods that match those selectors.
+The scheduler favors nodes that have fewer existing matching pods. Then, it schedules the pond on a node with the smallest number of
```
s/pond/pod

mburke5678 commented:
@ravisantoshgudimetla Which predicates are essential? Hard to tell from the code.

This PR lists essential predicates as: PodFitsResources, PodFitsHostPorts, HostName, MatchNodeSelector, which we don't list here.
kubernetes/kubernetes#50990
Are we listing only non-critical here?

ravisantoshgudimetla commented:
>  Are we listing only non-critical here?

General predicates are further split into non-critical and essential. Whatever we have currently are not under General predicates, so they are not under essential or non-critical. `PodFitsResources` comes under noncritical predicates.

mburke5678 commented:
@ravisantoshgudimetla I added more predicates from the Kubernetes source, including `PodFitsResources`. Are these the general predicates? If not, do you have a list? 

https://github.com/openshift/openshift-docs/pull/7313/files#diff-c881c052264120bfc3a506b42de92c15R234

mburke5678 commented:
@ravisantoshgudimetla @aveshagarwal Can you take a look to see if I have the predicates right?

mburke5678 commented:
@ravisantoshgudimetla @aveshagarwal  ^^

aveshagarwal commented:
@mburke5678 looking.

aveshagarwal commented:
PodFitsPorts (though still works) has been deprecated and is replaced with PodFitsHostPorts so would be good to mention this. So right now, both  PodFitsPorts and PodFitsHostPorts are same.

aveshagarwal commented:
Wonder why the `Available Predicates` does not mention about following default predicates: `NoVolumeZoneConflict`, `MaxEBSVolumeCount`, `MaxGCEPDVolumeCount`, `MaxAzureDiskVolumeCount`, `CheckNodeMemoryPressure`, `CheckNodeDiskPressure`, `CheckNodeCondition` and `PodToleratesNodeTaints`.

aveshagarwal commented:
NoVolumeNodeConflict does not have any description.

mburke5678 commented:
@ravisantoshgudimetla Thank you. I will look at your comments in the morning.

mburke5678 commented:
@ravisantoshgudimetla
_Wonder why the Available Predicates does not mention about following default predicates: NoVolumeZoneConflict, MaxEBSVolumeCount, MaxGCEPDVolumeCount, MaxAzureDiskVolumeCount, CheckNodeMemoryPressure, CheckNodeDiskPressure, CheckNodeCondition and PodToleratesNodeTaints, and CheckVolumeBinding, MatchInterPodAffinity._

These are further down under _Default Scheduler Policy_. Would it make more sense to put them under _Available Predicates > Static Predicates_?

aveshagarwal commented:
> These are further down under Default Scheduler Policy. Would it make more sense to put them under Available Predicates > Static Predicates

Yes.

aveshagarwal commented on admin_guide/scheduling/scheduler.adoc:
```diff
@@ -97,6 +101,105 @@ and a string match with the name of the host.
+{"name" : "HostName"}
+----
+
+**_PodFitsHostPorts_** Checks if a node has free ports for the requested pod ports.
```
Seems like you have described it twice and there are others too similarly. please fix.

aveshagarwal commented on admin_guide/scheduling/scheduler.adoc:
```diff
@@ -97,6 +101,105 @@ and a string match with the name of the host.
+{"name" : "PodFitsHostPorts"}
+----
+
+**_MatchNodeSelector_** Checks if a pod node selector matches the node label.
```
same as above.

aveshagarwal commented on admin_guide/scheduling/scheduler.adoc:
```diff
@@ -97,6 +101,105 @@ and a string match with the name of the host.
+{"name" : "MatchNodeSelector"}
+----
+
+**_PodFitsResources_** Checks if a node has sufficient resources, such as CPU, memory, GPU, opaque int resources, and so forth to run a pod.
```
same as above.

aveshagarwal commented on admin_guide/scheduling/scheduler.adoc:
```diff
@@ -97,6 +101,105 @@ and a string match with the name of the host.
+
+**_HostName_**
+----
+{"name" : "HostName"}
```
described twice.

aveshagarwal commented on admin_guide/scheduling/scheduler.adoc:
```diff
@@ -97,6 +101,105 @@ and a string match with the name of the host.
+----
+
+**_NoDiskConflict`_** Checks the volume requested by a pod is available.
+----
```
described twice.

aveshagarwal commented:
Default priority functions not described in this doc: InterPodAffinityPriority, NodePreferAvoidPodsPriority, NodeAffinityPriority, TaintTolerationPriority,

aveshagarwal commented:
Non-default priority functions: SelectorSpreadPriority, EqualPriority, ImageLocalityPriority, MostRequestedPriority, 

Non-default priority function behind alpha feature gate: ResourceLimitsPriority

mburke5678 commented:
@ravisantoshgudimetla @aveshagarwal 
Made Avesh's changes. 
**Can you help with definitions** for 
predicates: NoVolumeNodeConflict, CheckVolumeBinding, MaxAzureDiskVolumeCount
priorities: MostRequestedPriority, ImageLocalityPriority

mburke5678 commented:
@ravisantoshgudimetla @aveshagarwal Should we remove `NoVolumeNodeConflict`for 3,9 per:
https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG-1.9.md#storage-1 and https://bugzilla.redhat.com/show_bug.cgi?id=1540832

aveshagarwal commented:
> @ravisantoshgudimetla @aveshagarwal Should we remove NoVolumeNodeConflictfor 3,9 per:
https://github.com/kubernetes/kubernetes/blob/master/CHANGELOG-1.9.md#storage-1 and https://bugzilla.redhat.com/show_bug.cgi?id=1540832


Thanks for catching this. Yes I do not see NoVolumeNodeConflict so remove it.

aveshagarwal commented:
@mburke5678 sorry for delay. I am going to look into it soon.

aveshagarwal commented:
CheckVolumeBinding: 

`Evaluates if a pod can fit based on the volumes, it requests, for both bound and unbound PVCs. For PVCs that are bound, it checks that the corresponding PV's node affinity is satisfied by the given node. For PVCs that are unbound, it tries to find available PVs that can satisfy the PVC requirements and that the PV node affinity is satisfied by the given node. The predicate returns true if all bound PVCs have compatible PVs with the node, and if all unbound PVCs can be matched with an available and node-compatible PV.`

Please change as you need.

aveshagarwal commented:
MaxAzureDiskVolumeCount :+1: 
description similar to MaxGCEPDVolumeCount except it checks for Azure Disk Volumes.

aveshagarwal commented:
MostRequestedPriority:

`is a priority function that favors nodes with most requested resources. It calculates the percentage of memory and CPU requested by pods scheduled on the node, and prioritizes based on the maximum of the average of the fraction of requested to capacity.`

aveshagarwal commented:
ImageLocalityPriority:

`prioritizes nodes that already have requested pod container's images.`

aveshagarwal commented:
I think I have provided descriptions of predicates and priorities that you asked. Let me know if i missed something. I am still reviewing the doc though so might have some more comments.

aveshagarwal commented on admin_guide/scheduling/scheduler.adoc:
```diff
@@ -97,162 +307,345 @@ and a string match with the name of the host.
+----
+
+[[admin-guide-scheduler-general-predicates]]
+=== General Predicates
```
Not sure if by `General Predicates` here we mean in some literal sense? Otherwise, in scheduler, following are general predicates (divided into 2 categories, non-critical and essential): 
non-critical: PodFitsResources
essentials: PodFitsHostPorts, HostName, MatchNodeSelector

mburke5678 commented:
@qiancai @vikram-redhat assigned the BZ associated with this defect to Xiaowei Li who appears to no longer be with Red Hat. Can you reassign? 
I updated and reorganized the Default Scheduler topic [1], based on a [customer issue](https://access.redhat.com/support/cases/01980087). 
I updated the list of predicates and priorities to include all available. 
Also, I moved some sections around within the topic, and moved the _Controlling Pod Placement_ section into a new topic [2]. I did not change any wording in these sections. 
I removed the _Scheduler Extensibility_ sections, as they seemed redundant. 
Can you have someone on your team take a look? Please verify the predicates and priorities.  
[1] file:///home/mburke/Docs/openshift-docs/_preview/openshift-enterprise/BZ-1517586/admin_guide/scheduling/scheduler.html
[2] file:///home/mburke/Docs/openshift-docs/_preview/openshift-enterprise/BZ-1517586/admin_guide/scheduling/pod_placement.html#constraining-pod-placement-nodeselector

wjiangjay commented on admin_guide/scheduling/scheduler.adoc:
```diff
@@ -21,38 +21,208 @@ particular node.
+field in the
+xref:../../install_config/master_node_configuration.adoc#master-configuration-files[master configuration file].
+
+.Sample modified scheduler configuration file
```
yaml should be not supported here.

wjiangjay commented on admin_guide/scheduling/scheduler.adoc:
```diff
@@ -21,38 +21,208 @@ particular node.
+You can create a custom configuration, or use and modify one of the
+xref:scheduler-sample-policies[sample policy configurations].
+
+. Add any xref:configurable-predicates[configurable predicates] and xref:configurable-priorities[configurable-priority-functions]
```
for preview version, priority-function have wrong link, so please give a check here

wjiangjay commented on admin_guide/scheduling/scheduler.adoc:
```diff
@@ -97,185 +356,338 @@ and a string match with the name of the host.
+
+* If `presence` is false, and any of the requested labels match any of the
+node labels, it returns `false`. Otherwise, it returns `true`.
+* If `presence` is `true, and any of the requested labels do not match any of
```
true lack ` closed

wjiangjay commented on admin_guide/scheduling/scheduler.adoc:
```diff
@@ -97,185 +356,338 @@ and a string match with the name of the host.
+* If `presence` is false, and any of the requested labels match any of the
+node labels, it returns `false`. Otherwise, it returns `true`.
+* If `presence` is `true, and any of the requested labels do not match any of
 the node's labels, it returns false. Otherwise, it returns true.
```
Here false and true should be consistent with above items.

mburke5678 replied:
@wjiangjay @ravisantoshgudimetla 
What does this mean, _If `presence` is false_ and _If `presence` is true_? Does this mean if the labelsPresence parameter is present in the JSON, this is _true_ and if the parameter is absent, this is _false_?

mburke5678 commented:
@wjiangjay @ravisantoshgudimetla
What does this mean, If presence is false and If presence is true? Does this mean if the labelsPresence parameter is present in the JSON, this is true and if the parameter is absent, this is false?

mburke5678 commented:
@openshift/team-documentation PTAL. Not a XXL. Not all words need review. 
I updated and reorganized the Default Scheduler topic [1], based on a customer issue.
I updated the list of predicates and priorities to include all available.
Also, I moved some sections around within the topic, and moved the Controlling Pod Placement section into a new topic [2]. I did not change any wording in these sections.
I removed the Scheduler Extensibility sections, as they seemed redundant.
Can you have someone on your team take a look? Please verify the predicates and priorities.
[1] file:///home/mburke/Docs/openshift-docs/_preview/openshift-enterprise/BZ-1517586/admin_guide/scheduling/scheduler.html
[2] file:///home/mburke/Docs/openshift-docs/_preview/openshift-enterprise/BZ-1517586/admin_guide/scheduling/pod_placement.html#constraining-pod-placement-nodeselector

vikram-redhat commented:
@mburke5678 - those links are not correct? They are on your local file server?

mburke5678 commented:
@openshift/team-documentation 
Corrected links:
http://file.rdu.redhat.com/mburke/BZ-1517586/admin_guide/scheduling/scheduler.html
http://file.rdu.redhat.com/mburke/BZ-1517586/admin_guide/scheduling/pod_placement.html

ravisantoshgudimetla commented:
> , If presence is false and If presence is true? Does this mean if the labelsPresence parameter is present in the JSON

presence is true mean, we need to have the labels on the node.

mburke5678 commented:
@openshift/team-documentation PTAL 
Preview file:
http://file.rdu.redhat.com/mburke/BZ-1517586/admin_guide/scheduling/scheduler.html
This files is new, created from existing text in scheduler.html. Likely doesn't need much review; just to make sure I created the new file properly?
http://file.rdu.redhat.com/mburke/BZ-1517586/admin_guide/scheduling/pod_placement.html

adellape commented on admin_guide/scheduling/scheduler.adoc:
```diff
@@ -97,185 +393,350 @@ and a string match with the name of the host.
-These priority functions do not take any configuration parameters or inputs from
-the user. These are specified in the scheduler configuration using their exact
-name as well as the weight.
+=== Static priorities
```
"priorities" meant to be lowercase for sentence case, I take it?

adellape commented on admin_guide/scheduling/scheduler.adoc:
```diff
@@ -62,33 +269,122 @@ These predicates do not take any configuration parameters or inputs from the
+_The default scheduler policy includes the general predicates._
+
+
+**Non-critical general predicates**
```
Consider using `====` or `====` for this and the "Essential general predicates", with or without a leading `discrete`, instead of bold markup. I.e.:

```
[discrete]
==== Non-critical general predicates
```

adellape commented on admin_guide/scheduling/scheduler.adoc:
```diff
@@ -97,185 +393,350 @@ and a string match with the name of the host.
+the priority function has a weight of `1` except `*NodePreferAvoidPodsPriority*`,
+which has a weight of `10000`:
+
+**_SelectorSpreadPriority_** The scheduler looks for services, replication controllers (RC),
```
Could "The scheduler" be dropped here?

mburke5678 commented:
[rev_history]
|xref:../admin_guide/scheduling/scheduler.adoc#admin-guide-scheduler[Default Scheduling]
|Corrected the list of xref:../admin_guide/scheduling/scheduler.adoc#default-scheduler-policy[default scheduling policies and predicates]. Added definitions for all policies and predicates. Moved Controlling Pod Placement to new file. 
%
|xref:../admin_guide/scheduling/pod_placement.adoc#controlling-pod-placement[Controlling Pod Placement]
|New file with text moved from the _Default Scheduling_ topic.
%

mburke5678 commented:
[rev_history]
|xref:../admin_guide/scheduling/scheduler.adoc#admin-guide-scheduler[Default Scheduling]
|Corrected the list of xref:../admin_guide/scheduling/scheduler.adoc#default-scheduler-policy[default scheduling policies and predicates]. Added definitions for all policies and predicates. Moved Controlling Pod Placement to new file. 
%
|xref:../admin_guide/scheduling/pod_placement.adoc#controlling-pod-placement[Controlling Pod Placement]
|New file with text moved from the _Default Scheduling_ topic.
%

mburke5678 commented:
@wjiangjay Is the PR verified? I had a documentation peer review. Want to make sure you are OK with my changes before I merge.

mburke5678 commented:
/cherry-pick enterprise-3.6

mburke5678 commented:
/cherrypick enterprise-3.6

mburke5678 commented:
/cherrypick enterprise-3.6

mburke5678 commented:
/cherrypick enterprise-3.7

mburke5678 commented:
Cherry-pick
3.9 https://github.com/openshift/openshift-docs/pull/7914
3.7 https://github.com/openshift/openshift-docs/pull/7912
3/6 https://github.com/openshift/openshift-docs/pull/7908

vikram-redhat commented on architecture/additional_concepts/admission_controllers.adoc:
```diff
@@ -118,7 +118,7 @@ ifdef::openshift-enterprise,openshift-origin[]
 endif::[]
 ifdef::openshift-enterprise,openshift-origin[]
-- xref:../../admin_guide/scheduling/scheduler.adoc#controlling-pod-placement[Controlling Pod Placement]
+- xref:../../admin_guide/scheduling/pod_placement#controlling-pod-placement[Controlling Pod Placement]
```
@mburke5678 - missing .adoc causing build failures.

vikram-redhat commented on admin_guide/scheduling/pod_placement.adoc:
```diff
@@ -0,0 +1,380 @@
+endif::openshift-dedicated[]
+and xref:../../admin_guide/managing_projects.adoc#using-node-selectors[node selector]
+set up in your environment.
+//tag::node-selectors[]
```
@mburke5678 - because the tag moved from scheduling.adoc to this new file, any references to it needed to be moved as well. There was one reference in node_selector.adoc which was not moved.

PS: I am fixing the issues with this PR, this is just an FYI so you don't need to do anything.

vikram-redhat commented on admin_guide/scheduling/pod_placement.adoc:
```diff
@@ -0,0 +1,380 @@
+The `nodeSelectorLabelBlacklist` field of a master configuration file gives you
+control over the labels that certain roles can specify in a pod configuration's
+`nodeSelector` field. Users, service accounts, and groups that have the
+*pods/binding* permission xref:../../admin_guide/manage_rbac.adoc#admin-guide-manage-rbac[role]
```
@mburke5678 - manage_rbac.adoc was introduced in 3.7, not 3.6. Before that it was manage_users (from memory). This needed to point to that file instead of this one.

mburke5678 commented:
@ravisantoshgudimetla @aveshagarwal @wjiangjay 
Someone asked me to match the list of predicates to [openshift_master_facts_default_predicates.py](https://github.com/openshift/openshift-ansible/blob/master/roles/lib_utils/lookup_plugins/openshift_master_facts_default_predicates.py:). I cannot find the request. 
I changed the order of the predicates policies slightly to match and to groupd them into Default and Other predicates/policies.  I have a preview here. If no objections, I will merge this and make similar changes to 3.6 and 3.9
https://github.com/openshift/openshift-ansible/blob/master/roles/lib_utils/lookup_plugins/openshift_master_facts_default_predicates.py:

mburke5678 commented:
@aveshagarwal Sorry. I included the wrong link above (now fixed) The changes are on our preview server here:
http://file.rdu.redhat.com/mburke/enterprise-3.7/admin_guide/scheduling/scheduler.html#available-predicates
